book-content/chapters/04-essential-types-and-annotations.md

```diff
@@ -0,0 +1,2366 @@
+};
+
+// ---cut---
+const result = concatTwoStrings("Hello", "World");
+
+type test = Expect<Equal<typeof result, string>>;
+```
+
+Your job is to add some function paramater annotations to the `concatTwoStrings` function to make the errors go away.
+
+<Exercise title="Exercise 2: Annotating Empty Parameters" filePath="/src/015-essential-types-and-annotations/021-annotating-empty-parameters.problem.ts" resourceId="1fZdJK1AI9JNeRElmqAdKe"></Exercise>
+
+#### Exercise 3: The Basic Types
+
+As we've seen, TypeScript will show errors when types don't match.
+
+This set of examples shows us the basic types that TypeScript gives us to describe JavaScript:
+
+```ts twoslash
+// @errors: 2322
+export let example1: string = "Hello World!";
+export let example2: string = 42;
+export let example3: string = true;
+export let example4: string = Symbol();
+export let example5: string = 123n;
+```
+
+Note that the colon `:` is used to annotate the type of each variable, just like it was for typing the function parameters.
+
+You'll also notice there are several errors.
+
+Hovering over each of the underlined variables will display any associated error messages.
+
+For example, hovering over `example2` will show:
+
+```
+Type 'number' is not assignable to type 'string'.
+```
+
+The type error for `example3` tells us:
+
+```
+Type 'boolean' is not assignable to type 'string'.
+```
+
+Change the types of the annotations on each variable to make the errors go away.
+
+<Exercise title="Exercise 3: The Basic Types" filePath="/src/015-essential-types-and-annotations/022-all-types.problem.ts" resourceId="NMpTvrI4rUCyVa4GVzY1iN"></Exercise>
+
+#### Exercise 4: The `any` Type
+
+Here is a function called `handleFormData` that accepts an `e` typed as `any`. The function prevents the default form submission behavior, then creates an object from the form data and returns it:
+
+```ts
+const handleFormData = (e: any) => {
+  e.preventDefault();
+
+  const data = new FormData(e.terget);
+
+  const value = Object.fromEntries(data.entries());
+
+  return value;
+};
+```
+
+Here is a test for the function that creates a form, sets the `innerHTML` to add an input, and then manually submits the form. When it submits, we expect the value to equal the value that was in our form that we grafted in there:
+
+```ts
+it("Should handle a form submit", () => {
+  const form = document.createElement("form");
+
+  form.innerHTML = `
+<input name="name" value="John Doe"></Exercise>
+`;
+
+  form.onsubmit = (e) => {
+    const value = handleFormData(e);
+
+    expect(value).toEqual({ name: "John Doe" });
+  };
+
+  form.requestSubmit();
+
+  expect.assertions(1);
+});
+```
+
+Note that this isn't the normal way you would test a form, but it provides a way to test the example `handleFormData` function more extensively.
+
+In the code's current state, there are no red squiggly lines present.
+
+However, when running the test with Vitest we get an error similar to the following:
```

Comment:
U have included Vitest in the package.json deps in the Chapter 1 but other than that it's the first time it's mentioned and it might not be obvious how to run it etc.